File: python/day01.py
#!/usr/bin/env python

with open("../input/01") as f:
    depths = [int(l) for l in f]
    count = 0
    for i in range(1, len(depths)):
        if depths[i] > depths[i - 1]:
            count += 1
    print(f"Part 1: {count}")

    # Middle numbers of the sliding windows cancel out, so only need to compare the outer edges of the windows
    count = 0
    for i in range(3, len(depths)):
        if depths[i] > depths[i - 3]:
            count += 1
    print(f"Part 2: {count}")

File: rust/day01.rs
use std::fs;
use std::i32;
use std::str::FromStr;

fn main () {
    // Read input
    let filename = "../input/01";
    println!("In file {}", filename);

    let contents = fs::read_to_string(filename)
        .expect("Something went wrong reading the input file");

    let nums:Vec<i32> = contents.split_whitespace()
                        .map(|x| i32::from_str(x).unwrap())
                        .collect();

    // Part 1
    let mut count = 0;
    for i in 1..nums.len() {
        if nums[i] > nums[i-1] {
            count += 1;
        }
    }
    println!("Part 1: {}", count);

    // Part 2
    // Notice how the middle 2 numbers of the overlapping windows cancel out, so we only have to compare the outer edges
    count = 0;
    for i in 3..nums.len() {
        if nums[i] > nums[i-3] {
            count += 1;
        }
    }
    println!("Part 2: {}", count);
}